### DISCUSSION TITLE
umqtt.simple check_msg() error

### DISCUSSION THREAD

**Discussion by enginebeat** [micropython/micropython#11203] [Libraries & Drivers]
Hi All,
I'm using umqtt.simple to publish and subscribe to a topic as a test and can't get it to work.
The broker I'm using is the broker on Node-red Aedes. 

the publish part was working fine, but I can't get the subscribe part to work.

![image](https://user-images.githubusercontent.com/26185065/230035787-6e1e7cba-b340-41d8-a60d-753b0ab2f66c.png)

this is what I get on the debug window for node-red:
![image](https://user-images.githubusercontent.com/26185065/230036003-e1e906c7-08f6-4e4b-a9f6-b18d08067870.png)

code for the loop using check message:
` 
client = connectMQTT()

mySubscribe(client, "getCurrent")

count = 0
while(1):
        client.check_msg()
        #mypublish(client, "bat/current", str(getADCvalue(0)))
        utime.sleep(1)
        #count += 1`

code for the callback and the subscribe function:
```
def sub_callBack(topic, msg):
    print('received message %s on topic %s' % (msg, topic))
    #if(topic == "bat/getCurrent"):
        #if(msg == "0"):
            #mypublish(client, "bat/current", str(getADCvalue(0)))
    #data = [topic, msg]
    #return data

def mySubscribe(client, topic):
    client.set_callback(sub_callBack) # maybe this shouldn't be here...
    client.subscribe('getCurrent')
```

I know this is a bit messy, but this is my test code and I been modifying it to try to get it workin,  it should work. I omitted the connection part.
This is the error I get on the pico W side in Thonny:

> Traceback (most recent call last):
>   File "<stdin>", line 16, in <module>
>   File "/lib/umqtt/simple2.py", line 144, in check_msg
> TypeError: function takes 2 positional arguments but 4 were given

As far as I can see, I'm not doing anything that different from the examples I have seen, so I'm a bit stuck. I also have a subscriber on Node-red for the same topic and it works fine.

Any help would be appreciated.

Thank you.
Best Regards
               EngineBeat.

**Comment by andypiper** [comment]
I’m slightly confused about what you’re describing here. Which debug node is the node-red output from (and what is the full content of the message tree?). Which one is line 16 and/or line 144 in the MicroPython code? i.e. where are you providing too many arguments?

**Reply by enginebeat** [reply]
Hi Andy,
Node-red has a window on the left of the flows that will give info on what is happening called the debug window. line 16 is the line with client.check_msg() line 144 is on the lib/umqtt folder, file simple2.py.
your question is exactly my question, I'm not doing anything, as far as I know different from all the examples I have seen. I connect to the broker, set_callback, then subscribe to the topic and use check_msg().
it's check_message that is complaining.

Best Regards
                EB

**Comment by peterhinch** [comment]
I don't recognise `simple2.py`. The official code is [simple.py](https://github.com/micropython/micropython-lib/blob/master/micropython/umqtt.simple/umqtt/simple.py). We can't debug this unless we are discussing the same code :)

**Reply by enginebeat** [reply]
well, I tried to install simple.py but it wouldn't install, I seen on the list simple2.py, tried to install it and it installed  fine, so that's what I been using.
Best Regards
         EB

**Reply by andypiper** [reply]
what do you mean that it wouldn't install?

**Reply by enginebeat** [reply]
exactly that, I tried to install the libraries on my pico w with thonny and also with upip and it would not install.
simple2.py installed fine with thonny.

Best Regards
             EB

**Reply by ricksorensen** [reply]
References found with a quick search-
[pypi for simple2](https://pypi.org/project/micropython-umqtt.simple2/)
[github](https://github.com/fizista/micropython-umqtt.simple2)
Is initial install difficulty a mip/pypi conflict

**Comment by peterhinch** [comment]
The way to install official library modules is with [mip](http://docs.micropython.org/en/latest/reference/packages.html?highlight=mip#package-management).

It's hard for us to support unofficial modules: it requires us to study the documentation and code. In my view support queries should be directed to the module author. Unless, of course, the author has a presence here.

**Reply by peterhinch** [reply]
To add to this, I believe no official modules are hosted on PyPi.

To clarify my point on unofficial libraries - with official libraries there is a high probability that someone will offer support.

Unofficial libraries offer very useful facilities but it's worth considering if you need support and if so, how to get it. An experienced user will be happy to grapple with the code and offer fixes to any problems they find. If you might need support, before using one I would search here and in the old forum. Check whether the author responds to queries. If not, check whether others are using the package - if so, get a handle on what the general experience seems to be. 

If both searches draw a blank it may be time to look elsewhere.

**Comment by enginebeat** [comment]
Hi All,
Thank you or your help, I tried to install umqtt.simple again this morning, via Thonny's package manager and this time it installed, not quite sure why it didn't in previous attempts...
anyway, the program is now working.

I didn't really look at the origin of simple2, As it shows up as an option I assumed that it was an official release just a version 2 of the original umqtt.simple. will pay more attention to it next time.

Best Regards
             EB
